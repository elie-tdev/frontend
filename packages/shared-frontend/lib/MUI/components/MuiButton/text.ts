import type { ComponentsVariants } from '@mui/material/styles/variants'
import type { Theme } from '@mui/material/styles/createTheme'
import ifLight from '../../utils/if-mode'
import { HighlightColorsBase } from '../../utils/mixins'
import { colors } from './'

const textStyles = (color: HighlightColorsBase, theme: Theme) => {
  const { mixins } = theme
  return {
    boxShadow: '0 0 0px rgba(0,0,0,.001)',
    transition: 'color 300ms ease, box-shadow 300ms ease',
    '&:hover': {
      color: ifLight(
        t => t[color].dark,
        t => t[color].light,
        theme,
      ),
    },
    '&:focus': {
      color: ifLight(
        t => t[color].dark,
        t => t[color].light,
        theme,
      ),
    },
    '&:focus:hover': {
      color: ifLight(
        t => t[color].dark,
        t => t[color].light,
        theme,
      ),
    },
    '&.Mui-focusVisible': {
      color: ifLight(
        t => t[color].dark,
        t => t[color].light,
        theme,
      ),
      ...mixins._BAI_internal.bothOutline(color).focusVisible,
    },
    '&.MuiButtonGroup-groupedText:focus': {
      ...mixins._BAI_internal.groupText(color),
    },
    '&.MuiButtonGroup-groupedText:hover': {
      ...mixins._BAI_internal.groupText(color),
    },
    '&.MuiButtonGroup-groupedText:focus:hover': {
      ...mixins._BAI_internal.groupText(color),
    },
    '&.MuiButtonGroup-groupedText.Mui-focusVisible': {
      ...mixins._BAI_internal.groupText(color, true),
    },
  }
}

const text = (theme: Theme): NonNullable<ComponentsVariants['MuiButton']> =>
  colors.map(name => ({
    props: { variant: 'text', color: name },
    style: textStyles(name, theme),
  })) as NonNullable<ComponentsVariants['MuiButton']>

export default text
