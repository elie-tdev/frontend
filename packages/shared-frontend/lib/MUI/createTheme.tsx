import { createTheme as createMuiTheme } from '@mui/material/styles'

import { makeDefaultTheme } from './defaultTheme'
import { BAIThemeOptions } from './types'
import { createDefaultPaletteOptions } from './utils/palette'
import { createDefaultSpacingOptions } from './utils/spacing'

const createThemeBase = (themeOptions: BAIThemeOptions, isTest = false) => {
  createDefaultSpacingOptions(themeOptions)
  createDefaultPaletteOptions(themeOptions)
  return makeDefaultTheme(createMuiTheme(themeOptions), !isTest)
}

export const createTheme = (themeOptions: BAIThemeOptions = {}) =>
  createThemeBase(themeOptions)

export const createThemeTest = (themeOptions: BAIThemeOptions = {}) =>
  createThemeBase(themeOptions, true)
