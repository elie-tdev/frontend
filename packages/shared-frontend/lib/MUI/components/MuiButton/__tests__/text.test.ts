import { createThemeTest } from '../../../createTheme'
import ifLight from '../../../utils/if-mode'
import text from '../text'

describe('Text Button', () => {
  const theme = createThemeTest()
  const textButton = text(theme)

  it.each`
    key            | background
    ${'primary'}   | ${theme.palette.primary}
    ${'secondary'} | ${theme.palette.secondary}
    ${'success'}   | ${theme.palette.success}
    ${'error'}     | ${theme.palette.error}
    ${'warning'}   | ${theme.palette.warning}
    ${'info'}      | ${theme.palette.info}
  `(
    'should create styles for $key color',
    ({
      key,
      background,
    }: {
      key: string
      background: { main: string; dark: string; light: string }
    }) => {
      const variant = textButton.find(v => v.props.color === key)
      const textColor = (variant?.style as any)['&:hover'].color
      expect(textColor).toEqual(
        ifLight(
          _ => background.dark,
          _ => background.light,
          theme,
        ),
      )
      expect(variant).toMatchSnapshot()
    },
  )
})
