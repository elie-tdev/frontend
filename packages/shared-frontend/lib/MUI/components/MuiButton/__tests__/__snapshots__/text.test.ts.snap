// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Button should create styles for error color 1`] = `
Object {
  "props": Object {
    "color": "error",
    "variant": "text",
  },
  "style": Object {
    "&.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px #ef5350",
      "color": "#c62828",
    },
    "&.MuiButtonGroup-groupedText.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px rgba(239, 83, 80, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(239, 83, 80, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(239, 83, 80, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(239, 83, 80, 0.2)",
    },
    "&:focus": Object {
      "color": "#c62828",
    },
    "&:focus:hover": Object {
      "color": "#c62828",
    },
    "&:hover": Object {
      "color": "#c62828",
    },
    "boxShadow": "0 0 0px rgba(0,0,0,.001)",
    "transition": "color 300ms ease, box-shadow 300ms ease",
  },
}
`;

exports[`Text Button should create styles for info color 1`] = `
Object {
  "props": Object {
    "color": "info",
    "variant": "text",
  },
  "style": Object {
    "&.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px #03a9f4",
      "color": "#01579b",
    },
    "&.MuiButtonGroup-groupedText.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px rgba(3, 169, 244, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(3, 169, 244, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(3, 169, 244, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(3, 169, 244, 0.2)",
    },
    "&:focus": Object {
      "color": "#01579b",
    },
    "&:focus:hover": Object {
      "color": "#01579b",
    },
    "&:hover": Object {
      "color": "#01579b",
    },
    "boxShadow": "0 0 0px rgba(0,0,0,.001)",
    "transition": "color 300ms ease, box-shadow 300ms ease",
  },
}
`;

exports[`Text Button should create styles for primary color 1`] = `
Object {
  "props": Object {
    "color": "primary",
    "variant": "text",
  },
  "style": Object {
    "&.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px rgb(72, 145, 220)",
      "color": "rgb(18, 82, 148)",
    },
    "&.MuiButtonGroup-groupedText.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px rgba(72, 145, 220, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(72, 145, 220, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(72, 145, 220, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(72, 145, 220, 0.2)",
    },
    "&:focus": Object {
      "color": "rgb(18, 82, 148)",
    },
    "&:focus:hover": Object {
      "color": "rgb(18, 82, 148)",
    },
    "&:hover": Object {
      "color": "rgb(18, 82, 148)",
    },
    "boxShadow": "0 0 0px rgba(0,0,0,.001)",
    "transition": "color 300ms ease, box-shadow 300ms ease",
  },
}
`;

exports[`Text Button should create styles for secondary color 1`] = `
Object {
  "props": Object {
    "color": "secondary",
    "variant": "text",
  },
  "style": Object {
    "&.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px rgb(144, 144, 144)",
      "color": "rgb(81, 81, 81)",
    },
    "&.MuiButtonGroup-groupedText.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px rgba(144, 144, 144, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(144, 144, 144, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(144, 144, 144, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(144, 144, 144, 0.2)",
    },
    "&:focus": Object {
      "color": "rgb(81, 81, 81)",
    },
    "&:focus:hover": Object {
      "color": "rgb(81, 81, 81)",
    },
    "&:hover": Object {
      "color": "rgb(81, 81, 81)",
    },
    "boxShadow": "0 0 0px rgba(0,0,0,.001)",
    "transition": "color 300ms ease, box-shadow 300ms ease",
  },
}
`;

exports[`Text Button should create styles for success color 1`] = `
Object {
  "props": Object {
    "color": "success",
    "variant": "text",
  },
  "style": Object {
    "&.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px #4caf50",
      "color": "#1b5e20",
    },
    "&.MuiButtonGroup-groupedText.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px rgba(76, 175, 80, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(76, 175, 80, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(76, 175, 80, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(76, 175, 80, 0.2)",
    },
    "&:focus": Object {
      "color": "#1b5e20",
    },
    "&:focus:hover": Object {
      "color": "#1b5e20",
    },
    "&:hover": Object {
      "color": "#1b5e20",
    },
    "boxShadow": "0 0 0px rgba(0,0,0,.001)",
    "transition": "color 300ms ease, box-shadow 300ms ease",
  },
}
`;

exports[`Text Button should create styles for warning color 1`] = `
Object {
  "props": Object {
    "color": "warning",
    "variant": "text",
  },
  "style": Object {
    "&.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px #ff9800",
      "color": "#e65100",
    },
    "&.MuiButtonGroup-groupedText.Mui-focusVisible": Object {
      "boxShadow": "inset 0px 0px 0px 2px rgba(255, 152, 0, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(255, 152, 0, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:focus:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(255, 152, 0, 0.2)",
    },
    "&.MuiButtonGroup-groupedText:hover": Object {
      "boxShadow": "inset 0px 0px 0px 0px rgba(255, 152, 0, 0.2)",
    },
    "&:focus": Object {
      "color": "#e65100",
    },
    "&:focus:hover": Object {
      "color": "#e65100",
    },
    "&:hover": Object {
      "color": "#e65100",
    },
    "boxShadow": "0 0 0px rgba(0,0,0,.001)",
    "transition": "color 300ms ease, box-shadow 300ms ease",
  },
}
`;
